### ISSUE TITLE
[BUG] MaskedCMALLoss drops entire sequence on any single NaN timestep instead of elementwise masking

### ISSUE THREAD

**Issue by ashermetzger-rgb** [google-research/flood-forecasting#275]
**Location**
`googlehydrology/training/loss.py:403-408`

**Root Cause Mechanism**
Evaluating `mask = ~torch.isnan(ground_truth['y']).any(1).any(1)` collapses sequences into a 1D batch-level mask. If a single timestep in a sequence (e.g., 1 day out of 365) contains a missing value (`NaN`), the entire basin trajectory is dropped. When all sequences in a mini-batch contain at least one missing observation, `mask` evaluates to all-`False`, triggering empty tensor reductions and yielding `NaN` loss gradients.

**Impact**
Substantially degrades CMAL training efficiency on operational gauge datasets with intermittent missing observations by dropping valid discharge timesteps, while causing training crashes when batches contain no completely gap-free sequences.

**Suggested Fix**
Adopt pointwise boolean masking analogous to `MaskedMSELoss`, evaluating per-timestep log-likelihoods and normalizing strictly by the count of observed targets:

```python
valid_mask = ~torch.isnan(ground_truth['y'])  # shape: [batch, seq_len, target_dim]
y = torch.where(valid_mask, ground_truth['y'], torch.zeros_like(ground_truth['y']))

# ... compute mixture log_weights and log_like using sanitized y ...

result = torch.logsumexp(log_weights + log_like, dim=-1, keepdim=True)
valid_count = torch.clamp(valid_mask.sum(), min=1.0)

loss = -torch.sum(result * valid_mask.float()) / valid_count
return loss

```

**Verification Checklist**

* [ ] Test loss computation on synthetic batches with 10%, 50%, and 100% missing sequence entries.
* [ ] Check gradient propagation to ensure zero loss gradients flow from masked `NaN` positions.
* [ ] Validate that loss magnitude remains scale-invariant across batches with varying proportions of missing data.
